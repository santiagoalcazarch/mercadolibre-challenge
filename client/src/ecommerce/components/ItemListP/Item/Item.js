import React, { useMemo } from 'react'
import PropTypes from 'prop-types'
import Shipping from '../../../../shared/assets/icons/ic_shipping@2x.png'

/**
 * Muestra la informaciÃ³n de un item. Este item pertence
 * a la lista de items obtenida del API
 * @param {item} item
 * @returns 
 */
const Item = ({ item }) => {

  /**
   * Con los datos recibidos del API y utilizando Intl, se
   * renderiza adecuadamente el monto del item 
   */
  const itemPrice = useMemo(() => 
    new Intl.NumberFormat('es-CO', { 
      style: 'currency', 
      currency: item.price?.currency,
      maximumFractionDigits: 0,
      minimumFractionDigits: 0
    }).format( item.price?.amount )
  , [item.price]);
  
  return (
    <div className="item-container c-row">
      <div className="item-data col-sm-10">
        <div className="item-data-container d-flex align-items-top">
          <div className="item-image">
            <img src={item.picture} alt={item.title} />
          </div>
          <div className="item-data-desc ml-2 my-3">
            <div className="item-price-tx c-row">
              <span> { itemPrice } </span>
              {
                item.free_shipping && (
                  <div className="shipping-icon mx-3"> <img src={Shipping} alt="Free shipping" /> </div>
                )
              }
            </div>
            <div className="item-desc-tx">
              <span> { item.title } </span>
            </div>
          </div>
        </div>
      </div>
      <div className="item-location col-sm-2">
      { item.location }
      </div>
    </div>
  )
}

Item.propTypes = {
  item: PropTypes.object.isRequired
}

export default Item