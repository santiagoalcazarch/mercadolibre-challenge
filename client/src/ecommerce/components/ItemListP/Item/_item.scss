

.item-container{

  padding: 16px 0;

  .item-image{
    margin-right: 16px;
    text-align: center;

    img{
      width: 180px;
      height: 180px;
      object-fit: contain;
      border-radius: 4px;
      box-sizing: border-box;
      display: block;
    }
  }

  .item-data-desc{

    .item-price-tx{
      font-size: 24px;
      margin-bottom: 32px;

      .shipping-icon img{
        max-width: 20px;
      }
    }

    .item-desc-tx{
      font-size: 18px;
    }

  }

  .item-location{
    font-size: 12px;
    color: $mainGray;
    margin-top: 2rem;
  }
}