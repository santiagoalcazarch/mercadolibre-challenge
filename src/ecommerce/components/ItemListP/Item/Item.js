

import React from 'react'
import PropTypes from 'prop-types'
import imgTest from "./img.test.png";
import Shipping from '../../../../shared/assets/icons/ic_shipping@2x.png'

const Item = ({ item }) => {
  return (
    <div className="item-container c-row">
      <div className="item-data col-sm-10">
        <div className="item-data-container c-row align-items-top">
          <div className="item-image">
            <img src={imgTest}  />
          </div>
          <div className="item-data-desc ml-2 my-3">
            <div className="item-price-tx c-row">
              <div>$ 1.980</div>
              <div className="shipping-icon mx-3"> <img src={Shipping} alt="Free shipping" /> </div>
            </div>
            <div className="item-desc-tx">
              <span>Apple ipod touch 5g 16gb negro igual a nuev completo unico</span>
            </div>
          </div>
        </div>
      </div>
      <div className="item-location col-sm-2">
        Algun estado
      </div>
    </div>
  )
}

Item.propTypes = {}

export default Item