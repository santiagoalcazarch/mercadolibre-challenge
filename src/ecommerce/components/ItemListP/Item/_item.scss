

.item-container{

  padding: 16px 0;

  .item-image{
    margin-right: 16px;

    img{
      min-width: 180px;
      border-radius: 4px;
    }
  }

  .item-data-desc{

    .item-price-tx{
      font-size: 24px;
      margin-bottom: 32px;

      .shipping-icon img{
        max-width: 20px;
      }
    }

    .item-desc-tx{
      font-size: 18px;
    }

  }

  .item-location{
    font-size: 12px;
    color: $mainGray;
    margin-top: 2rem;
  }
}