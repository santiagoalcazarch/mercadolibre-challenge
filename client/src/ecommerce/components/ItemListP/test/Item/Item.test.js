
import { shallow } from "enzyme";
import Item from "../../Item/Item";

describe('Item component test', () => { 

  const item = {
    condition: "new",
    free_shipping: true,
    id: "MCO854837339",
    location: "Bogotá D.C.",
    picture: "http://http2.mlstatic.com/D_613152-MLA47073710736_082021-I.jpg",
    price: {currency: "COP", amount: 773500, decimals: 0},
    title: "Impresora de buena calidad"
  }

  const wrapper = shallow( <Item item={item} /> );

  test('Debería hacer match con el snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Debería desplegar el precio del item', () => {
    const itemPrice = new Intl.NumberFormat('es-CO', { 
      style: 'currency', 
      currency: item.price.currency,
      maximumFractionDigits: 0,
      minimumFractionDigits: 0
    }).format( item.price.amount );

    const priceText = wrapper.find('.item-price-tx span').text().trim();
    expect(priceText).toBe(itemPrice);
  })

  test('Debería desplegar la información del item correctamente', () => {

    const itemTitle = wrapper.find('.item-desc-tx span').text().trim();
    expect(itemTitle).toBe(item.title);

    const itemLocation = wrapper.find('.item-location').text().trim();
    expect(itemLocation).toBe(item.location);

    const freeShippingIcon = wrapper.find('.shipping-icon').exists();
    expect(freeShippingIcon).toBe(true);
  })

  test('Debería no desplegar la imagen de [free_shipping]', () => {

    const itemWithoutFreeShipping = { ...item, free_shipping: false }

    const wrapperFresS = shallow( <Item item={itemWithoutFreeShipping} /> );

    const freeShippingIcon = wrapperFresS.find('.shipping-icon').exists();

    expect(freeShippingIcon).toBe(false);
  })

})