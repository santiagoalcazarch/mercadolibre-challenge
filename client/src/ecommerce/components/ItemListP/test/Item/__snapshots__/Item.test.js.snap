// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item component test Debería hacer match con el snapshot 1`] = `
<div
  className="item-container c-row"
>
  <div
    className="item-data col-sm-10"
  >
    <div
      className="item-data-container d-flex align-items-top"
    >
      <div
        className="item-image"
      >
        <img
          alt="Impresora de buena calidad"
          src="http://http2.mlstatic.com/D_613152-MLA47073710736_082021-I.jpg"
        />
      </div>
      <div
        className="item-data-desc ml-2 my-3"
      >
        <div
          className="item-price-tx c-row"
        >
          <span>
             
            $ 773.500
             
          </span>
          <div
            className="shipping-icon mx-3"
          >
             
            <img
              alt="Free shipping"
              src="ic_shipping@2x.png"
            />
             
          </div>
        </div>
        <div
          className="item-desc-tx"
        >
          <span>
             
            Impresora de buena calidad
             
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="item-location col-sm-2"
  >
    Bogotá D.C.
  </div>
</div>
`;
